sr_unix/ydb_tp_s.c

```diff
@@ -0,0 +1,221 @@
+	tpfn_status = YDB_OK;
+	if (error_encountered)
+	{	/* If we reach here, it means an error occurred and "ydb_simpleapi_ch" was invoked which did a "longjmp"/"UNWIND" */
+		tpfn_status = TREF(ydb_error_code);
+		assert(0 < tpfn_status);
+		if (ERR_TPRETRY == tpfn_status)
+		{
+			assert(dollar_tlevel);	/* ensure "dollar_tlevel" is still non-zero */
+			/* If we reach here, it means we have a TPRETRY error from the database engine */
+			if (save_dollar_tlevel)
+			{	/* We were already inside a transaction when we entered this "ydb_tp_s" invocation.
+				 * So pass the TPRETRY to the caller until we go back to the outermost "ydb_tp_s" invocation.
+				 */
+				LIBYOTTADB_DONE;
+				REVERT;
+				assert(dollar_tlevel);
+				return YDB_TP_RESTART;
+			}
+			/* This is the outermost "ydb_tp_s" invocation. Do restart processing right here.
+			 * But before restarting the transaction cleanup database state of the partially completed transaction
+			 * using "preemptive_db_clnup". This is normally done by "mdb_condition_handler" in the case of YottaDB
+			 * runtime (since a TPRETRY is also an error) but that is not invoked by the simpleAPI and so we need this.
+			 * Note that if a restartable situation is detected inside of a call-in, it is possible the "tp_restart"
+			 * call happened right there. In that case, we need to skip the "tp_restart" call here. That is
+			 * detected by the "if" check below.
+			 */
+			assert((simpleapi_dollar_trestart == dollar_trestart)
+				|| (simpleapi_dollar_trestart == (dollar_trestart - 1)));
+			/* If the global "dollar_trestart" is different from what we noted down at the start of this try/retry
+			 * of this transaction ("simpleapi_dollar_trestart"), then it means we are done with "tp_restart" call
+			 * (since "dollar_trestart++" is done at the end of "tp_restart"). So use that as the basis of
+			 * the check to determine if a "tp_restart" is still needed or not.
+			 */
+			if (simpleapi_dollar_trestart == dollar_trestart)
+			{
+				preemptive_db_clnup(ERROR);	/* Cleanup "reset_gv_target", TREF(expand_prev_key) etc. */
+				if (cdb_sc_normal == t_fail_hist[t_tries])
+				{	/* User-induced TP restart. In that case, simulate TRESTART command in M.
+					 * This is relied upon by an assert in "tp_restart".
+					 */
+					op_trestart_set_cdb_code();
+				}
+				rc = tp_restart(1, !TP_RESTART_HANDLES_ERRORS);
+				assert((0 == rc) && (TPRESTART_STATE_NORMAL == tprestart_state));
+			}
+			tpfn_status = YDB_OK;	/* Now that the restart processing is complete, clear status back to normal */
+		}
+	}
+	LIBYOTTADB_DONE;		/* Shutoff active rtn indicator while call callback routine */
+	if (!save_dollar_tlevel)
+	{	/* This is the outermost TP. Note down "simpleapi_dollar_trestart" for later use
+		 * by "ydb_tp_s"/"ydb_simpleapi_ch" to help decide whether "tp_restart" processing is needed.
+		 */
+		simpleapi_dollar_trestart = dollar_trestart;
+	}
+	if (YDB_OK == tpfn_status)
```

Comment:
Would be nice to have a comment here that we are driving the transaction routine supplied by the user here.

Reply:
Done

Reply:
Yes